analyzers/golang/analyze.go

```diff
@@ -13,16 +15,86 @@ import (
 	"github.com/fossas/fossa-cli/buildtools/govendor"
 	"github.com/fossas/fossa-cli/buildtools/vndr"
 	"github.com/fossas/fossa-cli/errors"
+	"github.com/fossas/fossa-cli/files"
 	"github.com/fossas/fossa-cli/graph"
 	"github.com/fossas/fossa-cli/pkg"
 )
 
 // Analyze builds a dependency graph using go list and then looks up revisions
 // using tool-specific lockfiles.
 func (a *Analyzer) Analyze() (graph.Deps, error) {
+
+	// 1. Check for a set strategy.
+	switch a.Options.Strategy {
+	case "gomodules":
+		depGraph, err := gomodules.ModGraph("")
+		if len(depGraph.Direct) > 0 && err == nil {
+			return depGraph, nil
+		}
+		return gomodules.SumGraph("")
+	case "dep":
+		depGraph, err := dep.LockfileGraph("")
+		if len(depGraph.Direct) > 0 && err == nil {
+			return depGraph, nil
+		}
+		return dep.ManifestGraph("")
+	case "list":
+		return ListLockfileResolution(a)
+	}
+
+	// 2. Run `go list` and resolve using the lockfile.
+	depGraph, err := ListLockfileResolution(a)
+	if returnGraph(a.Module.Dir, err, len(depGraph.Direct)) {
+		return depGraph, nil
+	}
+
+	// 3. Check for gomodules.
+	if exists, err := files.Exists(filepath.Join(a.Module.Dir, "go.mod")); exists && err == nil {
+		depGraph, err := gomodules.ModGraph("")
+		if returnGraph(filepath.Join(a.Module.Dir, "go.mod"), err, len(depGraph.Direct)) {
+			return depGraph, nil
+		}
+	}
+	if exists, err := files.Exists(filepath.Join(a.Module.Dir, "go.sum")); exists && err == nil {
+		depGraph, err := gomodules.SumGraph("")
+		if returnGraph(filepath.Join(a.Module.Dir, "go.sum"), err, len(depGraph.Direct)) {
+			return depGraph, nil
+		}
+	}
+
+	// 4. Check for Dep.
+	if exists, err := files.Exists(filepath.Join(a.Module.Dir, "Gopkg.lock")); exists && err == nil {
+		depGraph, err := dep.LockfileGraph("")
+		if returnGraph(filepath.Join(a.Module.Dir, "Gopkg.lock"), err, len(depGraph.Direct)) {
+			return depGraph, nil
+		}
+	}
+	if exists, err := files.Exists(filepath.Join(a.Module.Dir, "Gopkg.toml")); exists && err == nil {
+		depGraph, err := dep.ManifestGraph("")
+		if returnGraph(filepath.Join(a.Module.Dir, "Gopkg.toml"), err, len(depGraph.Direct)) {
+			return depGraph, nil
+		}
+	}
+
+	return graph.Deps{}, err
+}
+
+func returnGraph(file string, err error, dependencyCount int) bool {
```

Comment:
@cnr I have found myself writing this function a lot. Lets somehow make it possible to only write this once in V2.

Reply:
agreed 👌